iscsi/iscsi.go

```diff
@@ -142,108 +158,101 @@ func getCurrentSessions() ([]iscsiSession, error) {
-			break
+		if fpath == nil {
+			return os.ErrNotExist
 		}
-		time.Sleep(time.Second * time.Duration(intervalSeconds))
+		// There might be a case that fpath contains multiple device paths if
+		// multiple PCI devices connect to same iscsi target. We handle this
+		// case at subsequent logic. Pick up only first path here.
+		*devicePath = fpath[0]
 	}
-	return false, err
+
+	return nil
 }
 
-func getMultipathDisk(path string) (string, error) {
-	// Follow link to destination directory
-	debug.Printf("Checking for multipath device for path: %s", path)
-	devicePath, err := os.Readlink(path)
-	if err != nil {
-		debug.Printf("Failed reading link for multipath disk: %s -- error: %s\n", path, err.Error())
-		return "", err
-	}
-	sdevice := filepath.Base(devicePath)
-	// If destination directory is already identified as a multipath device,
-	// just return its path
-	if strings.HasPrefix(sdevice, "dm-") {
-		debug.Printf("Already found multipath device: %s", sdevice)
-		return path, nil
-	}
-	// Fallback to iterating through all the entries under /sys/block/dm-* and
-	// check to see if any have an entry under /sys/block/dm-*/slaves matching
-	// the device the symlink was pointing at
-	dmPaths, err := filepath.Glob("/sys/block/dm-*")
-	if err != nil {
-		debug.Printf("Glob error: %s", err)
-		return "", err
-	}
-	for _, dmPath := range dmPaths {
-		sdevices, err := filepath.Glob(filepath.Join(dmPath, "slaves", "*"))
-		if err != nil {
-			debug.Printf("Glob error: %s", err)
-		}
-		for _, spath := range sdevices {
-			s := filepath.Base(spath)
-			debug.Printf("Basepath: %s", s)
-			if sdevice == s {
-				// We've found a matching entry, return the path for the
-				// dm-* device it was found under
-				p := filepath.Join("/dev", filepath.Base(dmPath))
-				debug.Printf("Found matching multipath device: %s under dm-* device path %s", sdevice, dmPath)
-				return p, nil
+// getMultipathDevice returns a multipath device for the configured targets if it exists
+func getMultipathDevice(devices []Device) (*Device, error) {
+	var multipathDevice *Device
+
+	for _, device := range devices {
+		if len(device.Children) != 1 {
+			multipathdNotRunning := ""
+			if len(device.Children) == 0 {
+				multipathdNotRunning = " (is multipathd running?)"
 			}
+			return nil, fmt.Errorf("device is not mapped to exactly one multipath device%s: %v", multipathdNotRunning, device.Children)
+		}
+		if multipathDevice != nil && device.Children[0].Name != multipathDevice.Name {
+			return nil, fmt.Errorf("devices don't share a common multipath device: %v", devices)
 		}
+		multipathDevice = &device.Children[0]
+	}
+
+	if multipathDevice == nil {
+		return nil, fmt.Errorf("multipath device not found")
 	}
-	debug.Printf("Couldn't find dm-* path for path: %s, found non dm-* path: %s", path, devicePath)
-	return "", fmt.Errorf("couldn't find dm-* path for path: %s, found non dm-* path: %s", path, devicePath)
+
+	if multipathDevice.Type != "mpath" {
```

Comment:
@paullaffitte it is good to have a tracker issue to check how this code block is effect if `user_friendly_names` is enabled in a multipath system. We can also add a todo here so that we wont forget to keep track of it.

Reply:
I think it will be better to have an issue for this as it make it more visible.

Reply:
I posted an issue about it: https://github.com/kubernetes-csi/csi-lib-iscsi/issues/32